Concurrent subscription to ReplaySubject leads to bogus onNext(NotificationLite) signals

As far as I know, the reactive streams specification imposes no restrictions on when or from which thread(s) a subscriber may subscribe to a source. However, if a subscriber subscribes to a `ReplaySubject` in parallel with the subject's upstream issuing a **legal** sequence of signals to that same `ReplaySubject`, the `ReplaySubject` can erroneously release an `onNext` signal carrying a `NotificationLite` value.

See the following repro:

```java
@Test
public void normal() throws Exception {
    for (int i = 1; i <= 1000000; i++) {
        Subject<String> source = ReplaySubject.create();
        Subject<String> sink = PublishSubject.create();
        TestObserver<String> observer = sink.test();
        Schedulers.computation().scheduleDirect(() -> {
            // issue signals to the source in adherence to the reactive streams specification
            source.onSubscribe(Disposable.empty());
            source.onNext("hello");
            source.onNext("world");
            source.onComplete();
        });
        Schedulers.computation().scheduleDirect(() -> {
            // connect the source to the sink in parallel with the signals issued to the source
            // note the cast() operator, which is here to detect non-String escapees
            source.cast(String.class).subscribe(sink);
        });
        observer.await().assertValues("hello", "world").assertComplete();
    }
}
```

When I run this, I get the following as output:

```
ReplaySubjectTest > normal() FAILED
    java.lang.AssertionError: Not completed (latch = 0, values = 2, errors = 1, completions = 0)
        at io.reactivex.rxjava3.observers.BaseTestConsumer.fail(BaseTestConsumer.java:128)
        at io.reactivex.rxjava3.observers.BaseTestConsumer.assertComplete(BaseTestConsumer.java:181)
        at joepembe.ReplaySubjectTest.normal(ReplaySubjectTest.java:57)

        Caused by:
        java.lang.ClassCastException: Cannot cast io.reactivex.rxjava3.internal.util.NotificationLite to java.lang.String
            at java.base/java.lang.Class.cast(Class.java:3606)
            at io.reactivex.rxjava3.internal.functions.Functions$CastToClass.apply(Functions.java:235)
            at io.reactivex.rxjava3.internal.operators.observable.ObservableMap$MapObserver.onNext(ObservableMap.java:58)
            at io.reactivex.rxjava3.subjects.ReplaySubject$UnboundedReplayBuffer.replay(ReplaySubject.java:791)
            at io.reactivex.rxjava3.subjects.ReplaySubject.subscribeActual(ReplaySubject.java:344)
            at io.reactivex.rxjava3.core.Observable.subscribe(Observable.java:13263)
            at io.reactivex.rxjava3.internal.operators.observable.ObservableMap.subscribeActual(ObservableMap.java:33)
            at io.reactivex.rxjava3.core.Observable.subscribe(Observable.java:13263)
            at joepembe.ReplaySubjectTest.lambda$normal$3(ReplaySubjectTest.java:55)
```

The same behavior occurs even when using `SerializedSubject` (not that I would expect anything different to occur, as `SerializedSubject` merely forwards the `subscribe()` invocation to the "real" subject):

```java
for (int i = 1; i <= 1000000; i++) {
    Subject<String> source = ReplaySubject.<String>create().toSerialized();
    Subject<String> sink = PublishSubject.<String>create().toSerialized();
    TestObserver<String> observer = sink.test();
    Schedulers.computation().scheduleDirect(() -> {
        // issue signals to the source in adherence to the reactive streams specification
        source.onSubscribe(Disposable.empty());
        source.onNext("hello");
        source.onNext("world");
        source.onComplete();
    });
    Schedulers.computation().scheduleDirect(() -> {
        // connect the source to the sink in parallel with the signals issued to the source
        // note the cast() operator, which is here to detect non-String escapees
        source.cast(String.class).subscribe(sink);
    });
    observer.await().assertValues("hello", "world").assertComplete();
}
```

Yields:

```
ReplaySubjectTest > serialized() FAILED
    java.lang.AssertionError: Not completed (latch = 0, values = 2, errors = 1, completions = 0)
        at io.reactivex.rxjava3.observers.BaseTestConsumer.fail(BaseTestConsumer.java:128)
        at io.reactivex.rxjava3.observers.BaseTestConsumer.assertComplete(BaseTestConsumer.java:181)
        at joepembe.ReplaySubjectTest.serialized(ReplaySubjectTest.java:33)

        Caused by:
        java.lang.ClassCastException: Cannot cast io.reactivex.rxjava3.internal.util.NotificationLite to java.lang.String
            at java.base/java.lang.Class.cast(Class.java:3606)
            at io.reactivex.rxjava3.internal.functions.Functions$CastToClass.apply(Functions.java:235)
            at io.reactivex.rxjava3.internal.operators.observable.ObservableMap$MapObserver.onNext(ObservableMap.java:58)
            at io.reactivex.rxjava3.subjects.ReplaySubject$UnboundedReplayBuffer.replay(ReplaySubject.java:791)
            at io.reactivex.rxjava3.subjects.ReplaySubject.subscribeActual(ReplaySubject.java:344)
            at io.reactivex.rxjava3.core.Observable.subscribe(Observable.java:13263)
            at io.reactivex.rxjava3.subjects.SerializedSubject.subscribeActual(SerializedSubject.java:49)
            at io.reactivex.rxjava3.core.Observable.subscribe(Observable.java:13263)
            at io.reactivex.rxjava3.internal.operators.observable.ObservableMap.subscribeActual(ObservableMap.java:33)
            at io.reactivex.rxjava3.core.Observable.subscribe(Observable.java:13263)
            at joepembe.ReplaySubjectTest.lambda$serialized$1(ReplaySubjectTest.java:31)
```

--- a/src/main/java/io/reactivex/rxjava3/subjects/ReplaySubject.java
+++ b/src/main/java/io/reactivex/rxjava3/subjects/ReplaySubject.java
@@ -652,8 +652,6 @@ public final class ReplaySubject<T> extends Subject<T> {
 
         final List<Object> buffer;
 
-        volatile boolean done;
-
         volatile int size;
 
         UnboundedReplayBuffer(int capacityHint) {
@@ -671,7 +669,6 @@ public final class ReplaySubject<T> extends Subject<T> {
             buffer.add(notificationLite);
             trimHead();
             size++;
-            done = true;
         }
 
         @Override
@@ -772,20 +769,17 @@ public final class ReplaySubject<T> extends Subject<T> {
 
                     Object o = b.get(index);
 
-                    if (done) {
-                        if (index + 1 == s) {
-                            s = size;
-                            if (index + 1 == s) {
-                                if (NotificationLite.isComplete(o)) {
-                                    a.onComplete();
-                                } else {
-                                    a.onError(NotificationLite.getError(o));
-                                }
-                                rs.index = null;
-                                rs.cancelled = true;
-                                return;
-                            }
-                        }
+                    if (NotificationLite.isComplete(o)) {
+                        a.onComplete();
+                        rs.index = null;
+                        rs.cancelled = true;
+                        return;
+                    } else
+                    if (NotificationLite.isError(o)) {
+                        a.onError(NotificationLite.getError(o));
+                        rs.index = null;
+                        rs.cancelled = true;
+                        return;
                     }
 
                     a.onNext((T)o);
@@ -856,8 +850,6 @@ public final class ReplaySubject<T> extends Subject<T> {
 
         Node<Object> tail;
 
-        volatile boolean done;
-
         SizeBoundReplayBuffer(int maxSize) {
             this.maxSize = maxSize;
             Node<Object> h = new Node<>(null);
@@ -895,7 +887,6 @@ public final class ReplaySubject<T> extends Subject<T> {
             t.lazySet(n); // releases both the tail and size
 
             trimHead();
-            done = true;
         }
 
         /**
@@ -1000,18 +991,17 @@ public final class ReplaySubject<T> extends Subject<T> {
 
                     Object o = n.value;
 
-                    if (done) {
-                        if (n.get() == null) {
-
-                            if (NotificationLite.isComplete(o)) {
-                                a.onComplete();
-                            } else {
-                                a.onError(NotificationLite.getError(o));
-                            }
-                            rs.index = null;
-                            rs.cancelled = true;
-                            return;
-                        }
+                    if (NotificationLite.isComplete(o)) {
+                        a.onComplete();
+                        rs.index = null;
+                        rs.cancelled = true;
+                        return;
+                    } else
+                    if (NotificationLite.isError(o)) {
+                        a.onError(NotificationLite.getError(o));
+                        rs.index = null;
+                        rs.cancelled = true;
+                        return;
                     }
 
                     a.onNext((T)o);
@@ -1068,8 +1058,6 @@ public final class ReplaySubject<T> extends Subject<T> {
         volatile TimedNode<Object> head;
 
         TimedNode<Object> tail;
-
-        volatile boolean done;
 
         SizeAndTimeBoundReplayBuffer(int maxSize, long maxAge, TimeUnit unit, Scheduler scheduler) {
             this.maxSize = maxSize;
@@ -1163,8 +1151,6 @@ public final class ReplaySubject<T> extends Subject<T> {
             size++;
             t.lazySet(n); // releases both the tail and size
             trimFinal();
-
-            done = true;
         }
 
         /**
@@ -1290,18 +1276,17 @@ public final class ReplaySubject<T> extends Subject<T> {
 
                     Object o = n.value;
 
-                    if (done) {
-                        if (n.get() == null) {
-
-                            if (NotificationLite.isComplete(o)) {
-                                a.onComplete();
-                            } else {
-                                a.onError(NotificationLite.getError(o));
-                            }
-                            rs.index = null;
-                            rs.cancelled = true;
-                            return;
-                        }
+                    if (NotificationLite.isComplete(o)) {
+                        a.onComplete();
+                        rs.index = null;
+                        rs.cancelled = true;
+                        return;
+                    } else
+                    if (NotificationLite.isError(o)) {
+                        a.onError(NotificationLite.getError(o));
+                        rs.index = null;
+                        rs.cancelled = true;
+                        return;
                     }
 
                     a.onNext((T)o);

--- a/src/test/java/io/reactivex/rxjava3/internal/operators/flowable/FlowableAmbTest.java
+++ b/src/test/java/io/reactivex/rxjava3/internal/operators/flowable/FlowableAmbTest.java
@@ -402,7 +402,8 @@ public class FlowableAmbTest extends RxJavaTest {
 
     @Test
     public void manySources() {
-        Flowable<?>[] a = new Flowable[32];
+        @SuppressWarnings("unchecked")
+        Flowable<Object>[] a = new Flowable[32];
         Arrays.fill(a, Flowable.never());
         a[31] = Flowable.just(1);
 

--- a/src/test/java/io/reactivex/rxjava3/internal/operators/observable/ObservableAmbTest.java
+++ b/src/test/java/io/reactivex/rxjava3/internal/operators/observable/ObservableAmbTest.java
@@ -235,7 +235,8 @@ public class ObservableAmbTest extends RxJavaTest {
 
     @Test
     public void manySources() {
-        Observable<?>[] a = new Observable[32];
+        @SuppressWarnings("unchecked")
+        Observable<Object>[] a = new Observable[32];
         Arrays.fill(a, Observable.never());
         a[31] = Observable.just(1);
 

--- a/src/test/java/io/reactivex/rxjava3/internal/operators/single/SingleAmbTest.java
+++ b/src/test/java/io/reactivex/rxjava3/internal/operators/single/SingleAmbTest.java
@@ -249,7 +249,8 @@ public class SingleAmbTest extends RxJavaTest {
 
     @Test
     public void manySources() {
-        Single<?>[] sources = new Single[32];
+        @SuppressWarnings("unchecked")
+        Single<Object>[] sources = new Single[32];
         Arrays.fill(sources, Single.never());
         sources[31] = Single.just(31);
 

--- a/src/test/java/io/reactivex/rxjava3/processors/ReplayProcessorTest.java
+++ b/src/test/java/io/reactivex/rxjava3/processors/ReplayProcessorTest.java
@@ -1832,4 +1832,69 @@ public class ReplayProcessorTest extends FlowableProcessorTest<Object> {
 
         rp.test().assertValuesOnly(4, 5);
     }
-}
+
+    @Test
+    public void terminationSubscriptionRaceUnbounded() throws Throwable {
+        for (int i = 1; i <= 10000; i++) {
+            ReplayProcessor<String> source = ReplayProcessor.create();
+            PublishProcessor<String> sink = PublishProcessor.create();
+            TestSubscriber<String> subscriber = sink.test();
+            Schedulers.computation().scheduleDirect(() -> {
+                // issue signals to the source in adherence to the reactive streams specification
+                source.onSubscribe(new BooleanSubscription());
+                source.onNext("hello");
+                source.onNext("world");
+                source.onComplete();
+            });
+            Schedulers.computation().scheduleDirect(() -> {
+                // connect the source to the sink in parallel with the signals issued to the source
+                // note the cast() operator, which is here to detect non-String escapees
+                source.cast(String.class).subscribe(sink);
+            });
+            subscriber.await().assertValues("hello", "world").assertComplete();
+        }
+    }
+
+    @Test
+    public void terminationSubscriptionRaceSizeBound() throws Throwable {
+        for (int i = 1; i <= 10000; i++) {
+            ReplayProcessor<String> source = ReplayProcessor.createWithSize(20);
+            PublishProcessor<String> sink = PublishProcessor.create();
+            TestSubscriber<String> subscriber = sink.test();
+            Schedulers.computation().scheduleDirect(() -> {
+                // issue signals to the source in adherence to the reactive streams specification
+                source.onSubscribe(new BooleanSubscription());
+                source.onNext("hello");
+                source.onNext("world");
+                source.onComplete();
+            });
+            Schedulers.computation().scheduleDirect(() -> {
+                // connect the source to the sink in parallel with the signals issued to the source
+                // note the cast() operator, which is here to detect non-String escapees
+                source.cast(String.class).subscribe(sink);
+            });
+            subscriber.await().assertValues("hello", "world").assertComplete();
+        }
+    }
+
+    @Test
+    public void terminationSubscriptionRaceTimeBound() throws Throwable {
+        for (int i = 1; i <= 10000; i++) {
+            ReplayProcessor<String> source = ReplayProcessor.createWithTime(20, TimeUnit.MINUTES, Schedulers.computation());
+            PublishProcessor<String> sink = PublishProcessor.create();
+            TestSubscriber<String> subscriber = sink.test();
+            Schedulers.computation().scheduleDirect(() -> {
+                // issue signals to the source in adherence to the reactive streams specification
+                source.onSubscribe(new BooleanSubscription());
+                source.onNext("hello");
+                source.onNext("world");
+                source.onComplete();
+            });
+            Schedulers.computation().scheduleDirect(() -> {
+                // connect the source to the sink in parallel with the signals issued to the source
+                // note the cast() operator, which is here to detect non-String escapees
+                source.cast(String.class).subscribe(sink);
+            });
+            subscriber.await().assertValues("hello", "world").assertComplete();
+        }
+    }}

--- a/src/test/java/io/reactivex/rxjava3/subjects/ReplaySubjectTest.java
+++ b/src/test/java/io/reactivex/rxjava3/subjects/ReplaySubjectTest.java
@@ -1378,4 +1378,70 @@ public class ReplaySubjectTest extends SubjectTest<Integer> {
 
         rs.test().assertValuesOnly(4, 5);
     }
+
+    @Test
+    public void terminationSubscriptionRaceUnbounded() throws Throwable {
+        for (int i = 1; i <= 10000; i++) {
+            Subject<String> source = ReplaySubject.create();
+            Subject<String> sink = PublishSubject.create();
+            TestObserver<String> observer = sink.test();
+            Schedulers.computation().scheduleDirect(() -> {
+                // issue signals to the source in adherence to the reactive streams specification
+                source.onSubscribe(Disposable.empty());
+                source.onNext("hello");
+                source.onNext("world");
+                source.onComplete();
+            });
+            Schedulers.computation().scheduleDirect(() -> {
+                // connect the source to the sink in parallel with the signals issued to the source
+                // note the cast() operator, which is here to detect non-String escapees
+                source.cast(String.class).subscribe(sink);
+            });
+            observer.await().assertValues("hello", "world").assertComplete();
+        }
+    }
+
+    @Test
+    public void terminationSubscriptionRaceSizeBound() throws Throwable {
+        for (int i = 1; i <= 10000; i++) {
+            Subject<String> source = ReplaySubject.createWithSize(20);
+            Subject<String> sink = PublishSubject.create();
+            TestObserver<String> observer = sink.test();
+            Schedulers.computation().scheduleDirect(() -> {
+                // issue signals to the source in adherence to the reactive streams specification
+                source.onSubscribe(Disposable.empty());
+                source.onNext("hello");
+                source.onNext("world");
+                source.onComplete();
+            });
+            Schedulers.computation().scheduleDirect(() -> {
+                // connect the source to the sink in parallel with the signals issued to the source
+                // note the cast() operator, which is here to detect non-String escapees
+                source.cast(String.class).subscribe(sink);
+            });
+            observer.await().assertValues("hello", "world").assertComplete();
+        }
+    }
+
+    @Test
+    public void terminationSubscriptionRaceTimeBound() throws Throwable {
+        for (int i = 1; i <= 10000; i++) {
+            Subject<String> source = ReplaySubject.createWithTime(20, TimeUnit.MINUTES, Schedulers.computation());
+            Subject<String> sink = PublishSubject.create();
+            TestObserver<String> observer = sink.test();
+            Schedulers.computation().scheduleDirect(() -> {
+                // issue signals to the source in adherence to the reactive streams specification
+                source.onSubscribe(Disposable.empty());
+                source.onNext("hello");
+                source.onNext("world");
+                source.onComplete();
+            });
+            Schedulers.computation().scheduleDirect(() -> {
+                // connect the source to the sink in parallel with the signals issued to the source
+                // note the cast() operator, which is here to detect non-String escapees
+                source.cast(String.class).subscribe(sink);
+            });
+            observer.await().assertValues("hello", "world").assertComplete();
+        }
+    }
 }

